### PR TITLE
L80: Add a proposal for ClientAsyncResponseReaderInterface destruction.

### PR DESCRIPTION
**Pull request by jacobsa**
See previous discussion in https://github.com/grpc/grpc/pull/26233.

@ctiller I've set you as the approver after the discussion in the thread above. Feel free to re-assign if that's not appropriate. Thanks!

### CODE REVIEW

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:20]: >This proposal provides a plan for fixing this so that other
> subclasses can count on their destructor running and their backing memory being
> freed, as with any other C++ class.

I would replace this with: 

```
This proposal provides a plan for guaranteed destructor invocation, such that 
subclasses can count on their destructor running and their backing memory being
freed, as with any other C++ class.
```

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:35]: nit: I would replace `introduced` with `included` to avoid repetition.

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:66]: > The interface's destructor is virtual, so other subclasses may have destructors
>  with side effects. However they are never actually run! ...

I would replace this with:

```
Even though subclasses can override the virtual destructor to add a side effect, their dtor does not actually run. This results in subtle bugs and confusing behavior, specially because   `delete p.release()` works differently than
 just letting a `std::unique_ptr p` go out of scope.
```

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:73]: nit: There are multiple hacks ...

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:83]: nit: I would reword this as: The plan is to land this change into steps

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:106]: This can break some random subclass that was relying on the dtor being a no-op.  I believe it needs a major gRPC version bump when we merge the commit.

We should clarify that here in the proposal.

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:120]: nit: as-is, but this requires ...

**Review by soheilhy — COMMENTED** [review]
Thank you for creating the gRFC.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md:106]: That is in theory also true of the change above; see the last paragraph. Does it need a major bump? Are major version bumps actually done, or does that make this a no-go?

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md]: Done.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md]: Done.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md]: Done.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md]: Done.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md]: Done (except I used `added`).

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md:20]: Done.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md]: Done.

**Comment by jacobsa** [comment]
@soheilhy Thanks for looking. I've addressed all your comments except the one I left open. PTAL.

(Sorry if I'm doing this wrong; I have no idea what conventions are for PRs, so am just treating this like Critique.)

**Comment by jacobsa** [comment]
Oh, I also set you to the approver in the doc. Also not sure what the process is there.

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [reply L79-cpp-async-response-reader-destruction.md:104]: We call "major version" bump as really going from gRPC 1.x to 1.x+1 not gRPC 2. Yes this is something that's done often and there is no issue doing so, once the gRFC is approved.

File: L79-cpp-async-response-reader-destruction.md
Comment (soheilhy) [inline L79-cpp-async-response-reader-destruction.md:8]: Before merging the PR, we usually send an email to the mailing list and then merge this PR.

**Review by soheilhy — APPROVED** [review]
Please add the group link before merging the PR.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md:104]: Okay, I added text saying each should result in a version bump. PTAL.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md:8]: Thanks. Do you want me to start the thread? Can you tell me how/give me an example?

**Review by soheilhy — CHANGES_REQUESTED** [review]
Sorry I am confused. I shouldn't have clicked approve.

We need two changes:
1) The title of the PR should change to include the suffix L79: ...
2) We need to post this PR to the group and post the thread link here for discussion.

File: L79-cpp-async-response-reader-destruction.md
Comment (jacobsa) [reply L79-cpp-async-response-reader-destruction.md]: Started the thread and left a link.

**Comment by jacobsa** [comment]
Thanks, both done.

**Comment by soheilhy** [comment]
> Thanks, both done.

All look great thanks!

**Comment by jacobsa** [comment]
@soheilhy: I believe the 10 business days timeout has been hit. Are you okay with merging this now?

**Review by ctiller — APPROVED** [review]

**Review by soheilhy — APPROVED** [review]

**Review by soheilhy — CHANGES_REQUESTED** [review]
@jacobsa Can you please update it to L80? Looks like L79 is already merged. Thanks!

**Comment by jacobsa** [comment]
Done; it's now L80. I don't know how to "resolve" this comment in github though. :-/

**Comment by jacobsa** [comment]
@soheilhy Sorry, meant to tag you before. As far as I can tell you need to clear the "requested changes" bit.

**Review by soheilhy — APPROVED** [review]
@jacobsa thanks for the update! yes, I have to click approve to make it ready to submit.

**Pull request merged**
